metagraph/src/graph/alignment/aligner_cigar.hpp

```diff
@@ -42,67 +45,65 @@ class Cigar {
     void append(Operator op, LengthType num = 1);
     void append(Cigar&& other);
 
-    void pop_front() {
-        assert(cigar_.size());
-        cigar_.erase(cigar_.begin(), cigar_.begin() + 1);
+    LengthType trim_clipping() {
+        if (cigar_.size() && cigar_.front().first == CLIPPED) {
+            LengthType ret_val = cigar_.front().second;
+            cigar_.erase(cigar_.begin(), cigar_.begin() + 1);
+            return ret_val;
+        } else {
+            return 0;
+        }
     }
 
-    void pop_back() {
-        assert(cigar_.size());
-        cigar_.pop_back();
+    LengthType trim_end_clipping() {
+        if (cigar_.size() && cigar_.back().first == CLIPPED) {
+            LengthType ret_val = cigar_.back().second;
+            cigar_.pop_back();
+            return ret_val;
+        } else {
+            return 0;
+        }
     }
 
-    typedef typename std::vector<value_type>::iterator iterator;
-    typedef typename std::vector<value_type>::const_iterator const_iterator;
+    LengthType get_clipping() const {
+        return cigar_.size() && cigar_.front().first == CLIPPED ? cigar_.front().second : 0;
+    }
 
-    // This is essentially just a vector, so there's no reason not to have it editable
-    iterator begin() { return cigar_.begin(); }
-    iterator end() { return cigar_.end(); }
-    const_iterator begin() const { return cigar_.cbegin(); }
-    const_iterator end() const { return cigar_.cend(); }
+    LengthType get_end_clipping() const {
+        return cigar_.size() && cigar_.back().first == CLIPPED ? cigar_.back().second : 0;
+    }
 
-    template <typename... Args>
-    void insert(iterator it, Args&&... args) {
-        cigar_.insert(it, std::forward<Args>(args)...);
+    void extend_clipping(LengthType n) {
+        assert(cigar_.size());
+        if (cigar_.front().first != CLIPPED) {
+            cigar_.insert(begin(), value_type(CLIPPED, n));
+        } else {
+            cigar_.front().second += n;
+        }
     }
 
-    value_type& front() { return cigar_.front(); }
-    value_type& back() { return cigar_.back(); }
-    const value_type& front() const { return cigar_.front(); }
-    const value_type& back() const { return cigar_.back(); }
+    const_iterator begin() const { return cigar_.cbegin(); }
+    const_iterator end() const { return cigar_.cend(); }
```

Comment:
If you want to use range-iteration, why not write `auto x : cigar.data()`? So these begin/end would be redundant.

Reply:
yes, but these iterators only add four extra lines, so it's not adding that much redundancy...

Reply:
I know, but there are still no benefits of having them, imo. It just adds another way to do the same thing, which is generally not a good thing.